File: devices.go
package main

import (
	"math"

	. "github.com/linux-surface/iptsd/processing"
	"github.com/pkg/errors"
)

type IptsStylusDevice struct {
	Serial    uint32
	Device    *UinputDevice
	Processor *StylusProcessor
}

type IptsTouchDevice struct {
	Device    *UinputDevice
	Processor *TouchProcessor
}

type IptsDevices struct {
	Touch *IptsTouchDevice

	ActiveStylus *IptsStylusDevice
	Styli        []*IptsStylusDevice
}

func IptsDevicesGetRes(virt int, phys int) int32 {
	res := float64(virt*10) / float64(phys)
	return int32(math.Round(res))
}

func IptsDevicesCreateStylus(ipts *IptsContext) (*UinputDevice, error) {
	dev := &UinputDevice{
		Name:    "IPTS Stylus",
		Vendor:  ipts.DeviceInfo.Vendor,
		Product: ipts.DeviceInfo.Product,
		Version: uint16(ipts.DeviceInfo.Version),
	}

	err := dev.Open()
	if err != nil {
		return nil, err
	}

	dev.SetEvbit(EV_KEY)
	dev.SetEvbit(EV_ABS)

	dev.SetPropbit(INPUT_PROP_DIRECT)
	dev.SetPropbit(INPUT_PROP_POINTER)

	dev.SetKeybit(BTN_TOUCH)
	dev.SetKeybit(BTN_STYLUS)
	dev.SetKeybit(BTN_TOOL_PEN)
	dev.SetKeybit(BTN_TOOL_RUBBER)

	dev.SetAbsInfo(ABS_X, UinputAbsInfo{
		Minimum:    0,
		Maximum:    9600,
		Resolution: IptsDevicesGetRes(9600, ipts.Config.Width),
	})

	dev.SetAbsInfo(ABS_Y, UinputAbsInfo{
		Minimum:    0,
		Maximum:    7200,
		Resolution: IptsDevicesGetRes(7200, ipts.Config.Height),
	})

	dev.SetAbsInfo(ABS_PRESSURE, UinputAbsInfo{
		Minimum: 0,
		Maximum: 4096,
	})

	dev.SetAbsInfo(ABS_TILT_X, UinputAbsInfo{
		Minimum:    -9000,
		Maximum:    9000,
		Resolution: 5730,
	})

	dev.SetAbsInfo(ABS_TILT_Y, UinputAbsInfo{
		Minimum:    -9000,
		Maximum:    9000,
		Resolution: 5730,
	})

	dev.SetAbsInfo(ABS_MISC, UinputAbsInfo{
		Minimum: 0,
		Maximum: 65535,
	})

	err = dev.Create()
	if err != nil {
		return nil, err
	}

	return dev, nil
}

func IptsDevicesCreateTouch(ipts *IptsContext) (*UinputDevice, error) {
	dev := &UinputDevice{
		Name:    "IPTS Touch",
		Vendor:  ipts.DeviceInfo.Vendor,
		Product: ipts.DeviceInfo.Product,
		Version: uint16(ipts.DeviceInfo.Version),
	}

	err := dev.Open()
	if err != nil {
		return nil, err
	}

	dev.SetEvbit(EV_ABS)
	dev.SetPropbit(INPUT_PROP_DIRECT)

	dev.SetAbsInfo(ABS_MT_SLOT, UinputAbsInfo{
		Minimum: 0,
		Maximum: int32(ipts.DeviceInfo.MaxTouchPoints),
	})

	dev.SetAbsInfo(ABS_MT_TRACKING_ID, UinputAbsInfo{
		Minimum: 0,
		Maximum: int32(ipts.DeviceInfo.MaxTouchPoints),
	})

	dev.SetAbsInfo(ABS_MT_POSITION_X, UinputAbsInfo{
		Minimum:    0,
		Maximum:    9600,
		Resolution: IptsDevicesGetRes(9600, ipts.Config.Width),
	})

	dev.SetAbsInfo(ABS_MT_POSITION_Y, UinputAbsInfo{
		Minimum:    0,
		Maximum:    7200,
		Resolution: IptsDevicesGetRes(7200, ipts.Config.Height),
	})

	err = dev.Create()
	if err != nil {
		return nil, err
	}

	return dev, nil
}

func (devices *IptsDevices) AddStylus(ipts *IptsContext, serial uint32) error {
	stylus, err := IptsDevicesCreateStylus(ipts)
	if err != nil {
		return err
	}

	processor := &StylusProcessor{}
	processor.Flush()

	devices.ActiveStylus = &IptsStylusDevice{
		Device:    stylus,
		Serial:    serial,
		Processor: processor,
	}
	devices.Styli = append(devices.Styli, devices.ActiveStylus)

	return nil
}

func (devices *IptsDevices) Create(ipts *IptsContext) error {
	if ipts.Config.Width == 0 || ipts.Config.Height == 0 {
		return errors.Errorf("Display size is 0")
	}

	touch, err := IptsDevicesCreateTouch(ipts)
	if err != nil {
		return err
	}

	processor := &TouchProcessor{
		InvertX:        ipts.Config.InvertX,
		InvertY:        ipts.Config.InvertY,
		MaxTouchPoints: int(ipts.DeviceInfo.MaxTouchPoints),
	}

	devices.Touch = &IptsTouchDevice{
		Device:    touch,
		Processor: processor,
	}

	devices.AddStylus(ipts, uint32(0))

	return nil
}

func (devices *IptsDevices) Destroy() error {
	err := devices.Touch.Device.Destroy()
	if err != nil {
		return err
	}

	for _, stylus := range devices.Styli {
		err = stylus.Device.Destroy()
		if err != nil {
			return err
		}
	}

	return nil
}

File: devices.c
// SPDX-License-Identifier: GPL-2.0-or-later

#include <errno.h>
#include <fcntl.h>
#include <linux/uinput.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#include "devices.h"
#include "stylus-processing.h"
#include "syscall.h"
#include "touch-processing.h"

static int iptsd_devices_res(int virt, int phys)
{
	double res = (double)(virt * 10) / (double)phys;
	return (int)roundf(res);
}

static int iptsd_devices_create_stylus(struct iptsd_stylus_device *stylus,
		struct iptsd_device_config config)
{
	struct uinput_setup setup;
	struct uinput_abs_setup abs_setup;

	int file = iptsd_syscall_open("/dev/uinput", O_WRONLY | O_NONBLOCK);
	if (file < 0)
		return file;

	stylus->dev = file;
	memset(&setup, 0, sizeof(struct uinput_setup));
	memset(&abs_setup, 0, sizeof(struct uinput_abs_setup));

	iptsd_syscall_ioctl(file, UI_SET_EVBIT, (void *)EV_KEY);
	iptsd_syscall_ioctl(file, UI_SET_EVBIT, (void *)EV_ABS);

	iptsd_syscall_ioctl(file, UI_SET_PROPBIT, (void *)INPUT_PROP_DIRECT);
	iptsd_syscall_ioctl(file, UI_SET_PROPBIT, (void *)INPUT_PROP_POINTER);

	iptsd_syscall_ioctl(file, UI_SET_KEYBIT, (void *)BTN_TOUCH);
	iptsd_syscall_ioctl(file, UI_SET_KEYBIT, (void *)BTN_STYLUS);
	iptsd_syscall_ioctl(file, UI_SET_KEYBIT, (void *)BTN_TOOL_PEN);
	iptsd_syscall_ioctl(file, UI_SET_KEYBIT, (void *)BTN_TOOL_RUBBER);

	abs_setup.code = ABS_X;
	abs_setup.absinfo.minimum = 0;
	abs_setup.absinfo.maximum = 9600;
	abs_setup.absinfo.resolution = iptsd_devices_res(9600, config.width);
	iptsd_syscall_ioctl(file, UI_ABS_SETUP, &abs_setup);

	abs_setup.code = ABS_Y;
	abs_setup.absinfo.minimum = 0;
	abs_setup.absinfo.maximum = 7200;
	abs_setup.absinfo.resolution = iptsd_devices_res(7200, config.height);
	iptsd_syscall_ioctl(file, UI_ABS_SETUP, &abs_setup);

	abs_setup.code = ABS_PRESSURE;
	abs_setup.absinfo.minimum = 0;
	abs_setup.absinfo.maximum = 4096;
	abs_setup.absinfo.resolution = 0;
	iptsd_syscall_ioctl(file, UI_ABS_SETUP, &abs_setup);

	abs_setup.code = ABS_TILT_X;
	abs_setup.absinfo.minimum = -9000;
	abs_setup.absinfo.maximum = 9000;
	abs_setup.absinfo.resolution = 5730;
	iptsd_syscall_ioctl(file, UI_ABS_SETUP, &abs_setup);

	abs_setup.code = ABS_TILT_Y;
	abs_setup.absinfo.minimum = -9000;
	abs_setup.absinfo.maximum = 9000;
	abs_setup.absinfo.resolution = 5730;
	iptsd_syscall_ioctl(file, UI_ABS_SETUP, &abs_setup);

	abs_setup.code = ABS_MISC;
	abs_setup.absinfo.minimum = 0;
	abs_setup.absinfo.maximum = 65535;
	abs_setup.absinfo.resolution = 0;
	iptsd_syscall_ioctl(file, UI_ABS_SETUP, &abs_setup);

	setup.id.bustype = BUS_VIRTUAL;
	setup.id.vendor = config.vendor;
	setup.id.product = config.product;
	setup.id.version = config.version;
	strcpy(setup.name, "IPTS Stylus");

	int ret = iptsd_syscall_ioctl(file, UI_DEV_SETUP, &setup);
	if (ret < 0)
		return ret;

	return iptsd_syscall_ioctl(file, UI_DEV_CREATE, NULL);
}

static int iptsd_devices_create_touch(struct iptsd_touch_device *touch,
		struct iptsd_device_config config)
{
	struct uinput_setup setup;
	struct uinput_abs_setup abs_setup;

	int file = iptsd_syscall_open("/dev/uinput", O_WRONLY | O_NONBLOCK);
	if (file < 0)
		return file;

	touch->dev = file;
	touch->processor.max_contacts = config.max_contacts;
	touch->processor.invert_x = config.invert_x;
	touch->processor.invert_y = config.invert_y;

	int ret = iptsd_touch_processing_init(&touch->processor);
	if (ret < 0)
		return ret;

	memset(&setup, 0, sizeof(struct uinput_setup));
	memset(&abs_setup, 0, sizeof(struct uinput_abs_setup));

	iptsd_syscall_ioctl(file, UI_SET_EVBIT, (void *)EV_ABS);
	iptsd_syscall_ioctl(file, UI_SET_PROPBIT, (void *)INPUT_PROP_DIRECT);

	abs_setup.code = ABS_MT_SLOT;
	abs_setup.absinfo.minimum = 0;
	abs_setup.absinfo.maximum = config.max_contacts;
	abs_setup.absinfo.resolution = 0;
	iptsd_syscall_ioctl(file, UI_ABS_SETUP, &abs_setup);

	abs_setup.code = ABS_MT_TRACKING_ID;
	abs_setup.absinfo.minimum = 0;
	abs_setup.absinfo.maximum = config.max_contacts;
	abs_setup.absinfo.resolution = 0;
	iptsd_syscall_ioctl(file, UI_ABS_SETUP, &abs_setup);

	abs_setup.code = ABS_MT_POSITION_X;
	abs_setup.absinfo.minimum = 0;
	abs_setup.absinfo.maximum = 9600;
	abs_setup.absinfo.resolution = iptsd_devices_res(9600, config.width);
	iptsd_syscall_ioctl(file, UI_ABS_SETUP, &abs_setup);

	abs_setup.code = ABS_MT_POSITION_Y;
	abs_setup.absinfo.minimum = 0;
	abs_setup.absinfo.maximum = 7200;
	abs_setup.absinfo.resolution = iptsd_devices_res(7200, config.height);
	iptsd_syscall_ioctl(file, UI_ABS_SETUP, &abs_setup);

	setup.id.bustype = BUS_VIRTUAL;
	setup.id.vendor = config.vendor;
	setup.id.product = config.product;
	setup.id.version = config.version;
	strcpy(setup.name, "IPTS Touch");

	ret = iptsd_syscall_ioctl(file, UI_DEV_SETUP, &setup);
	if (ret < 0)
		return ret;

	return iptsd_syscall_ioctl(file, UI_DEV_CREATE, NULL);
}

int iptsd_devices_add_stylus(struct iptsd_devices *devices, uint32_t serial)
{
	struct iptsd_stylus_device *stylus = NULL;

	for (int i = 0; i < IPTSD_MAX_STYLI; i++) {
		if (devices->styli[i].active)
			continue;

		stylus = &devices->styli[i];
	}

	if (!stylus)
		return 0;

	stylus->active = true;
	stylus->serial = serial;
	devices->active_stylus = stylus;

	iptsd_stylus_processing_flush(&stylus->processor);

	return iptsd_devices_create_stylus(stylus, devices->config);
}

int iptsd_devices_emit(int fd, int type, int code, int val)
{
	struct input_event ie;

	memset(&ie, 0, sizeof(struct input_event));

	ie.type = type;
	ie.code = code;
	ie.value = val;

	return iptsd_syscall_write(fd, &ie, sizeof(struct input_event));
}

int iptsd_devices_create(struct iptsd_devices *devices)
{
	if (devices->config.width == 0 || devices->config.height == 0) {
		fprintf(stderr, "Display size is 0!\n");
		return -EINVAL;
	}

	int ret = iptsd_devices_create_touch(&devices->touch, devices->config);
	if (ret < 0) {
		fprintf(stderr, "Failed to create touch device: %s\n",
				iptsd_syscall_strerr(ret));
		return ret;
	}

	ret = iptsd_devices_add_stylus(devices, 0);
	if (ret < 0) {
		fprintf(stderr, "Failed to create stylus: %s\n",
				iptsd_syscall_strerr(ret));
		return ret;
	}

	return 0;
}

void iptsd_devices_destroy(struct iptsd_devices *devices)
{
	iptsd_syscall_ioctl(devices->touch.dev, UI_DEV_DESTROY, NULL);
	iptsd_syscall_close(devices->touch.dev);

	iptsd_touch_processing_free(&devices->touch.processor);

	for (int i = 0; i < IPTSD_MAX_STYLI; i++) {
		struct iptsd_stylus_device stylus = devices->styli[i];

		if (!stylus.active)
			continue;

		iptsd_syscall_ioctl(stylus.dev, UI_DEV_DESTROY, NULL);
		iptsd_syscall_close(stylus.dev);
	}
}

